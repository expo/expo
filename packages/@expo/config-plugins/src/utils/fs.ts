import fs from 'fs';
import path from 'path';

/** A basic function that copies a single file to another file location. */
export async function copyFilePathToPathAsync(src: string, dest: string): Promise<void> {
  const srcFile = await fs.promises.readFile(src);
  await fs.promises.mkdir(path.dirname(dest), { recursive: true });
  await fs.promises.writeFile(dest, srcFile);
}

/** Remove a single file (not directory). Returns `true` if a file was actually deleted. */
export function removeFile(filePath: string): boolean {
  try {
    fs.unlinkSync(filePath);
    return true;
  } catch (error: any) {
    // Skip if the remove did nothing.
    if (error.code === 'ENOENT') {
      return false;
    }
    throw error;
  }
}
