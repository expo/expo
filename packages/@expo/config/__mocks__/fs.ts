module.exports = require('memfs').fs;

// NOTE(cedric): workaround to also mock `node:fs`
jest.mock('node:fs', () => require('fs'));
