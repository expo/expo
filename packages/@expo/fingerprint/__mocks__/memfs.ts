const memfs = jest.requireActual('memfs');
const path = require('path');

/**
 * Further mock memfs `fs.promises` to ensure the parent directory exists.
 * We used to call `fs.promises.mkdtemp(path.join(os.tmpdir(), ...))`, so that we don't have to worry about os.tmpdir() exists.
 */
const origMkdtemp = memfs.fs.promises.mkdtemp;
memfs.fs.promises.mkdtemp = (
  prefix: string,
  options?: Parameters<typeof origMkdtemp>[1]
): ReturnType<typeof origMkdtemp> => {
  memfs.fs.mkdirSync(path.dirname(prefix), { recursive: true });
  return origMkdtemp(prefix, options);
};

module.exports = memfs;
