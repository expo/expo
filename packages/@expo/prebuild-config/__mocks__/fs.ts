const fs = require('memfs').fs;

// Add fs.promises.cp polyfill for memfs since it doesn't support it
fs.promises.cp = async (src: string, dest: string, options?: { recursive?: boolean }) => {
  const path = require('path');

  const srcStat = await fs.promises.stat(src);

  if (srcStat.isDirectory()) {
    await fs.promises.mkdir(dest, { recursive: true });
    const entries = await fs.promises.readdir(src, { withFileTypes: true });

    for (const entry of entries) {
      const srcPath = path.join(src, entry.name);
      const destPath = path.join(dest, entry.name);

      if (entry.isDirectory()) {
        await fs.promises.cp(srcPath, destPath, options);
      } else {
        await fs.promises.copyFile(srcPath, destPath);
      }
    }
  } else {
    await fs.promises.copyFile(src, dest);
  }
};

module.exports = fs;

jest.mock('node:fs', () => require('memfs').fs);
