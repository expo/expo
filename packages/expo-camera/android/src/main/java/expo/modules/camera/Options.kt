package expo.modules.camera

import android.media.CamcorderProfile
import expo.modules.kotlin.records.Field
import expo.modules.kotlin.records.Record

class PictureOptions : Record {
  @Field val quality: Double = 1.0
  @Field val base64: Boolean = false
  @Field val exif: Boolean = false
  @Field val additionalExif: Map<String, Any>? = null
  @Field val skipProcessing: Boolean = false
  @Field val fastMode: Boolean = false
  @Field val id: Int? = null
  @Field val maxDownsampling: Int = 1
}

class RecordingOptions : Record {
  @Field val maxDuration: Int = -1
  @Field val maxFileSize: Int = -1
  @Field val quality: Int = CamcorderProfile.QUALITY_HIGH
  @Field val mute: Boolean = false
  @Field val videoBitrate: Int? = null
}
