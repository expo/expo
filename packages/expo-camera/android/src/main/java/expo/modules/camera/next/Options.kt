package expo.modules.camera.next

import expo.modules.camera.next.records.VideoQuality
import expo.modules.kotlin.records.Field
import expo.modules.kotlin.records.Record

data class PictureOptions(
  @Field val quality: Double = 1.0,
  @Field val base64: Boolean = false,
  @Field val exif: Boolean = false,
  @Field val additionalExif: Map<String, Any>? = null,
  @Field val skipProcessing: Boolean = false,
  @Field val fastMode: Boolean = false,
  @Field val id: Int? = null,
  @Field val maxDownsampling: Int = 1
) : Record

data class RecordingOptions(
  @Field val maxDuration: Int = 0,
  @Field val maxFileSize: Int = 0,
  @Field val quality: VideoQuality?
) : Record
