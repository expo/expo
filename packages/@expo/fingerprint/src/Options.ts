import fs from 'fs/promises';
import os from 'os';
import path from 'path';

import type { NormalizedOptions, Options } from './Fingerprint.types';

export const FINGERPRINT_IGNORE_FILENAME = '.fingerprintignore';

export const DEFAULT_IGNORE_PATHS = [
  FINGERPRINT_IGNORE_FILENAME,
  '**/android/build/**/*',
  '**/android/app/build/**/*',
  '**/android/app/.cxx/**/*',
  '**/ios/Pods/**/*',

  // Ignore all expo configs because we will read expo config in a HashSourceContents already
  'app.config.ts',
  'app.config.js',
  'app.config.json',
  'app.json',
];

export async function normalizeOptionsAsync(
  projectRoot: string,
  options?: Options
): Promise<NormalizedOptions> {
  return {
    ...options,
    platforms: options?.platforms ?? ['android', 'ios'],
    concurrentIoLimit: options?.concurrentIoLimit ?? os.cpus().length,
    hashAlgorithm: options?.hashAlgorithm ?? 'sha1',
    ignorePaths: await collectIgnorePathsAsync(projectRoot, options),
  };
}

async function collectIgnorePathsAsync(projectRoot: string, options?: Options): Promise<string[]> {
  const ignorePaths = [
    ...DEFAULT_IGNORE_PATHS,
    ...(options?.ignorePaths ?? []),
    ...(options?.dirExcludes?.map((dirExclude) => `${dirExclude}/**/*`) ?? []),
  ];

  const fingerprintIgnorePath = path.join(projectRoot, FINGERPRINT_IGNORE_FILENAME);
  try {
    const fingerprintIgnore = await fs.readFile(fingerprintIgnorePath, 'utf8');
    const fingerprintIgnoreLines = fingerprintIgnore.split('\n');
    for (const line of fingerprintIgnoreLines) {
      const trimmedLine = line.trim();
      if (trimmedLine) {
        ignorePaths.push(trimmedLine);
      }
    }
  } catch {}

  return ignorePaths;
}
