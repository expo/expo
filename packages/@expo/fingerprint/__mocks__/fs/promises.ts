import { fs } from 'memfs';
import path from 'path';

/**
 * Further mock `fs.promises` to ensure the parent directory exists.
 * We used to call `fs.promises.mkdtemp(path.join(os.tmpdir(), ...))`, so that we don't have to worry about os.tmpdir() exists.
 */
const origMkdtemp = fs.promises.mkdtemp;
fs.promises.mkdtemp = (
  prefix: string,
  options?: Parameters<typeof origMkdtemp>[1]
): ReturnType<typeof origMkdtemp> => {
  fs.mkdirSync(path.dirname(prefix), { recursive: true });
  return origMkdtemp(prefix, options);
};

module.exports = fs.promises;
