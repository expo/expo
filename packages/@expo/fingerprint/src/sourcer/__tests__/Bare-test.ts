import spawnAsync from '@expo/spawn-async';
import fs from 'fs';
import { vol } from 'memfs';
import path from 'path';

import { normalizeOptionsAsync } from '../../Options';
import {
  getBareAndroidSourcesAsync,
  getBareIosSourcesAsync,
  getRncliAutolinkingSourcesAsync,
} from '../Bare';

jest.mock('@expo/spawn-async');
jest.mock('fs/promises');
jest.mock('/app/package.json', () => ({}), { virtual: true });

describe('getBareSourcesAsync', () => {
  afterEach(() => {
    vol.reset();
  });

  it('should contain android and ios folders in bare react-native project', async () => {
    vol.fromJSON(require('./fixtures/BareReactNative70Project.json'));
    let sources = await getBareAndroidSourcesAsync('/app', await normalizeOptionsAsync('/app'));
    expect(sources).toContainEqual(expect.objectContaining({ filePath: 'android', type: 'dir' }));

    sources = await getBareIosSourcesAsync('/app', await normalizeOptionsAsync('/app'));
    expect(sources).toContainEqual(expect.objectContaining({ filePath: 'ios', type: 'dir' }));
  });
});

describe(getRncliAutolinkingSourcesAsync, () => {
  afterEach(() => {
    vol.reset();
  });

  it('should contain rn-cli autolinking projects', async () => {
    const mockSpawnAsync = spawnAsync as jest.MockedFunction<typeof spawnAsync>;
    const fixture = fs.readFileSync(
      path.join(__dirname, 'fixtures', 'RncliAutoLinking.json'),
      'utf8'
    );
    mockSpawnAsync.mockResolvedValue({
      stdout: fixture,
      stderr: '',
      status: 0,
      signal: null,
      output: [fixture, ''],
    });
    const sources = await getRncliAutolinkingSourcesAsync(
      '/root/apps/demo',
      await normalizeOptionsAsync('/app')
    );
    expect(sources).toContainEqual(
      expect.objectContaining({
        type: 'dir',
        filePath: '../../node_modules/react-native-reanimated',
      })
    );
    expect(sources).toMatchSnapshot();
  });

  it('should not contain absolute paths', async () => {
    const mockSpawnAsync = spawnAsync as jest.MockedFunction<typeof spawnAsync>;
    const fixture = fs.readFileSync(
      path.join(__dirname, 'fixtures', 'RncliAutoLinking.json'),
      'utf8'
    );
    mockSpawnAsync.mockResolvedValue({
      stdout: fixture,
      stderr: '',
      status: 0,
      signal: null,
      output: [fixture, ''],
    });
    const sources = await getRncliAutolinkingSourcesAsync(
      '/root/apps/demo',
      await normalizeOptionsAsync('/app')
    );
    for (const source of sources) {
      if (source.type === 'dir' || source.type === 'file') {
        expect(source.filePath).not.toMatch(/^\/root/);
      } else {
        expect(source.contents).not.toMatch(/"\/root\//);
      }
    }
  });
});
