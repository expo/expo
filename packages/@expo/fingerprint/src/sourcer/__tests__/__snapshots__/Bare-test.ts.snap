// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getRncliAutolinkingSourcesAsync should contain rn-cli autolinking projects 1`] = `
[
  {
    "filePath": "../../node_modules/react-native-reanimated",
    "reasons": [
      "bareRncliAutolinking",
    ],
    "type": "dir",
  },
  {
    "contents": "{"react-native-reanimated":{"root":"../../node_modules/react-native-reanimated","name":"react-native-reanimated","platforms":{"ios":{"podspecPath":"../../node_modules/react-native-reanimated/RNReanimated.podspec","configurations":[],"scriptPhases":[]},"android":{"sourceDir":"../../node_modules/react-native-reanimated/android","packageImportPath":"import com.swmansion.reanimated.ReanimatedPackage;","packageInstance":"new ReanimatedPackage()","buildTypes":[],"componentDescriptors":[],"androidMkPath":"../../node_modules/react-native-reanimated/android/build/generated/source/codegen/jni/Android.mk","cmakeListsPath":"../../node_modules/react-native-reanimated/android/build/generated/source/codegen/jni/CMakeLists.txt"}}}}",
    "id": "rncliAutolinkingConfig",
    "reasons": [
      "bareRncliAutolinking",
    ],
    "type": "contents",
  },
]
`;
